### PR TITLE
#12 MkBucket.list() must show directories

### PR DESCRIPTION
**Pull request by super132**
For task #12 , implemented directories listing functionality in `MkBucket.list()`

### CODE REVIEW

**Comment by super132** [comment]
@dmarkov , please review this PR. Thanks.

**Comment by dmarkov** [comment]
@super132 Thanks for your pull request, let me find someone who can review it

**Comment by dmarkov** [comment]
@pinaf please review, thanks

**Comment by dmarkov** [comment]
> @dmarkov , please review this PR. Thanks.

@super132 thanks, will take a look

File: src/main/java/com/jcabi/s3/Bucket.java [hunk 1]
```diff
@@ -183,9 +183,18 @@ public void remove(final String key) throws IOException {
                     new Function<String, String>() {
                         @Override
                         public String apply(final String input) {
-                            return input.substring(
-                                Bucket.Prefixed.this.prefix.length()
-                            );
+                            final String result;
+                            if ((input.length()
+                                < Bucket.Prefixed.this.prefix.length())
```
Comment (pinaf) [inline src/main/java/com/jcabi/s3/Bucket.java]: @super132 readibility here is bad. I suggest a final local variable 
```java
final String prefix = Bucket.Prefixed.this.prefix
```

File: src/main/java/com/jcabi/s3/Bucket.java [hunk 2]
```diff
@@ -183,9 +183,18 @@ public void remove(final String key) throws IOException {
...
+                                && (Bucket.Prefixed.this.prefix.contains(input))
```
Comment (pinaf) [inline src/main/java/com/jcabi/s3/Bucket.java]: @super132 we can use the new var here

File: src/main/java/com/jcabi/s3/Bucket.java [hunk 3]
```diff
@@ -183,9 +183,18 @@ public void remove(final String key) throws IOException {
...
+                            ) {
+                                result = "/";
+                            } else {
+                                result = input.substring(
+                                    Bucket.Prefixed.this.prefix.length()
```
Comment (pinaf) [inline src/main/java/com/jcabi/s3/Bucket.java]: @super132 and here

**Comment by pinaf** [comment]
@super132 3 comments above

**Comment by super132** [comment]
@pinaf I have addressed the comments. Please take a look. Thanks.

**Comment by pinaf** [comment]
@rultor merge

File: src/main/java/com/jcabi/s3/Bucket.java [hunk 4]
```diff
@@ -183,9 +183,16 @@ public void remove(final String key) throws IOException {
...
+                            final String result;
+                            final String head = Bucket.Prefixed.this.prefix;
+                            if ((input.length() < head.length())
+                                && (head.contains(input))
+                            ) {
+                                result = "/";
+                            } else {
+                                result = input.substring(head.length());
+                            }
+                            return result;
```
Comment (yegor256) [inline src/main/java/com/jcabi/s3/Bucket.java]: @super132 I didn't understand this code at all... why did you change it if the ticket is about `MkBucket`?

**Comment by yegor256** [comment]
@super132 one question from me above

File: src/main/java/com/jcabi/s3/Bucket.java (hunk 4 above)
Comment (super132) [inline src/main/java/com/jcabi/s3/Bucket.java]: @yegor256 that is because I found that if `Bucket.Prefixed` used in the test is not capable to show up root directory with prefix if `MKBucket` return so. Should I basically get rid of `Bucket.Prefixed` usage in the test and create a new ticket on `Bucket.Prefixed` or keeping this change? Thanks.

**Comment by super132** [comment]
@yegor256 I have replied o your comment. Thanks.

**Comment by yegor256** [comment]
@super132 changing that `Bucket.Prefixed` class is not a good idea. You better change the tests

**Comment by super132** [comment]
@yegor256 I have reverted the changes in `Bucket` and modified the test case to use different prefix. Would you please take a look? Thanks.

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by dmarkov** [comment]
@pinaf **17 mins** added to your account (payment number `AP-221248548Y708474E`), many thanks for your contribution! 120 hours and 17 mins spent here.. review comments (c=2) added as a bonus. +17 to your rating, your total score is [+8113](http://www.netbout.com/b/35128?open=rating)

**Comment by dmarkov** [comment]
@rultor please deploy
